pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -772,7 +772,16 @@ func (npm *nominatedPodMap) delete(p *v1.Pod) {
 }
 
 func (npm *nominatedPodMap) update(oldPod, newPod *v1.Pod) {
-	// We update irrespective of the nominatedNodeName changed or not, to ensure
+	// In some cases, an Update event with no "NominatedNode" present is received right
+	// after a node("NominatedNode") is reserved for this pod in memory.
+	// If we go updating (delete and add) it, it actually un-reserves the node since
+	// the newPod doesn't carry sped.status.nominatedNode.
+	// In this case, during this time other low-priority pods have chances to take space which
+	// was reserved for the nominatedPod.
+	if len(oldPod.Status.NominatedNodeName) == 0 && len(newPod.Status.NominatedNodeName) == 0 {
```

Comment:
What you have found is an interesting case. The only problem is that a pod pointer could change in an update. If we bypass this update, we may not receive the new pointer.

Reply:
That's true. The good thing is that in both npm.delete() and npm.add(), we actually check whether pod.UID exists or not, rather than directly check the pointer's presence. And the idea of this change skips only one case which is both old and new pods don't have NominatedNodeName set.

Reply:
I understand, but with the existing logic, if a pod pointer is updated, this update function updates the pod pointer. This condition that you have added skips the pointer update. Your point is valid though that this skips only when both NominatedNodeNames are empty, which can be true only for pods that have just been nominated and we have not received their NominatedNode status update.
This may solve the problem, but it does not look great and feels like a hack to me. Let me think about a bit more and see if I can find a cleaner solution.

Reply:
That's fair. We can brainstorm for a better solution.

Reply:
@bsalamat A new solution is proposed. PTAL.